tape/Ctape_dmpfil.man
.\" Copyright (C) 1990-2003 by CERN/IT/PDP/DM
.\" All rights reserved
.\"
.TH CTAPE_DMPFIL "3castor" "$Date: 2003/10/14 07:17:16 $" CASTOR "Ctape Library Functions"
.SH NAME
Ctape_dmpfil \- analyze the content of a tape
.SH SYNOPSIS
.B #include <sys/types.h>
.br
\fB#include "Ctape_api.h"\fR
.sp
.BI "int Ctape_dmpinit (char *" path ,
.BI "char *" vid ,
.BI "char *" density ,
.BI "char *" drive ,
.BI "char *" devtype ,
.BI "int " maxblksize ,
.BI "int " fromblock ,
.BI "int " toblock ,
.BI "int " maxbyte ,
.BI "int " fromfile ,
.BI "int " maxfile ,
.BI "int " code ,
.BI "int " flags );
.sp
.BI "int Ctape_dmpfil (char *" path ,
.BI "char *" lbltype ,
.BI "int *" blksize,
.BI "char *" fid,
.BI "int *" fsec,
.BI "int *" fseq,
.BI "int *" lrecl,
.BI "char *" recfm,
.BI "u_signed64 *" Size );
.sp
.BI "int Ctape_dmpend ();
.SH DESCRIPTION
.B Ctape_dmpfil
analyzes the content of a tape. Each call analyzes one file and
gives summary information like label type, record format, actual block size
and file size. Detailed information and actual block content is reported using
the
.B Ctape_dmpmsg
routine.
The application should call
.B Ctape_mount
to mount the tape,
.B Ctape_dmpinit
to initialize the dump routine, loop on
.B Ctape_dmpfil
and finally call
.BR Ctape_dmpend .
.TP
.I path
specifies the pathname relative to the current directory or the full pathname
used in
.BR Ctape_mount .
.TP
.I vid
is the volume visual identifier.
It must be at most six characters long and is converted to upper case.
.TP
.I density
The supported alphanumeric densities are:
.br
.B 1600
or
.B 6250
for round (IBM 3420) tapes,
.B 38000
for standard IBM 3480 cartridges,
.B 38KC
for 3480 cartridges with compaction (IDRC),
.B 38KD
for double density IBM 3490 cartridges,
.B 38KDC
for 3490 cartridges with compaction (IDRC),
.BR 8200 ,
.BR 8200C ,
.B 8500
or
.B 8500C
for Exabytes,
.B DDS
or
.B DDSC
for DAT cartridges,
.BR 2G ,
.BR 6G ,
.BR 10G ,
.BR 10GC ,
.BR 20G ,
.BR 20GC ,
.BR 35G ,
.BR 35GC ,
.B 40G
or
.B 40GC
for DLT cartridges,
.B 10G
or
.B 10GC
for IBM 3590 cartridges,
.B 40G
or
.B 40GC
for IBM 3590E cartridges,
.BR 10G ,
.BR 10GC ,
.BR 25G ,
.BR 25GC ,
.B 50G
or
.B 50GC
for Redwood cartridges,
.B 20G
or
.B 20GC
for STK 9840 cassettes,
.B 60G
or
.B 60GC
for STK 9940 cartridges,
.B 100G
or
.B 100GC
for LTO cartridges.
.TP
.I drive
specifies the drive name.
.TP
.I maxblksize
specifies the maximum block size. Default is 65536.
.TP
.I fromblock
Block number of the first block to be analyzed in each file.
Default is 1.
.TP
.I toblock
Block number of the last block to be analyzed in each file.
Default is
.IR fromblock .
.TP
.I maxbyte
Number of bytes per block to be dumped (0 means all bytes).
Default is 320.
.TP
.I fromfile
specifies that the content of the tape should be analyzed starting at this
file number. Default 1.
.TP
.I maxfile
Number of files to be dumped (0 means all files; this the default
for 3420/3480/3490).
The default value for long media (Exabytes, DLTs, IBM3590) is one data file.
.TP
.I code
Character code for data interpretation.
It can be set to
.B DMP_ASC
or
.BR DMP_EBC .
.TP
.I flags
If set to
.BR IGNOREEOI ,
Ctape_dmpfil will not take two consecutive tapemarks as End Of Information.
If
.I maxfile
is zero,
the dump will terminate only when encountering blank tape or too many parity
errors (on IBM3420, the reel would wind off).
.SH RETURN VALUE
These routines return 0 if the operation was successful or -1 if the operation
failed. In the latter case,
.B serrno
is set appropriately.
.B Ctape_dmpfil
returns also 1 when the dump is complete.
.SH ERRORS
.TP 1.2i
.B EIO
I/O error.
.TP
.B ENXIO
Drive not operational.
.TP
.B EINVAL
Bad parameter value.
.TP
.B ETBLANK
Blank tape.
.TP
.B ETCOMPA
Drive/cartridge compatibility problem.
.TP
.B ETHWERR
Device malfunction.
.TP
.B ETPARIT
Parity error.
.TP
.B ETUNREC
Unrecoverable media error.
.TP
.B ETNOSNS
No sense data available.
.SH SEE ALSO
.BR Ctape_mount(3)
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


tape/Ctape_dmpfil.c
/*
 * Copyright (C) 1990-2002 by CERN/IT/PDP/DM
 * All rights reserved
 */


/*	Ctape_dmpfil - analyse the content of a tape file */
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <errno.h>
#include <unistd.h>
#include "h/Ctape.h"
#include "h/Ctape_api.h"
#include "h/serrno.h"
#include "h/u64subr.h"
void (*Ctape_dmpmsg) (int, const char *, ...) = NULL;
static char *buffer;
static struct {
	char vid[CA_MAXVIDLEN+1];
	char devtype[CA_MAXDVTLEN+1];
	int maxblksize;
	int fromblock;
	int toblock;
	int maxbyte;
	int fromfile;
	int maxfile;
	int code;
	int flags;
} dmpparm;
static int infd = -1;
static int irec;

int Ctape_dmpinit(char *path,
                  char *vid,
                  char *aden,
                  char *drive,
                  char *devtype,
                  int maxblksize,
                  int fromblock,
                  int toblock,
                  int maxbyte,
                  int fromfile,
                  int maxfile,
                  int code,
                  int flags)
{
	int errflg = 0;

	if (! vid || *vid == '\0') {
		Ctape_dmpmsg (MSG_ERR, TP031);
		errflg++;
	}
	if (code != DMP_ASC && code != DMP_EBC) {
		Ctape_dmpmsg (MSG_ERR, TP006, "code");
		errflg++;
	}
	if (flags != 0 && flags != IGNOREEOI) {
		Ctape_dmpmsg (MSG_ERR, TP006, "flags");
		errflg++;
	}
	if (errflg) {
		serrno = EINVAL;
		return (-1);
	}

	Ctape_dmpmsg (MSG_OUT, " DUMP - TAPE MOUNTED ON DRIVE %s\n", drive);

	if (maxblksize < 0)
		maxblksize = DMP_DEFAULTBLOCKSIZE;
	if (maxbyte < 0) maxbyte = 320;
	if (maxfile < 0) {
			maxfile = 1;
        }
	if (fromblock < 0) fromblock = 1;
	if (toblock < 0) toblock = 1;

	buffer = malloc (maxblksize);

	/* print headers */

	Ctape_dmpmsg (MSG_OUT, " DUMP - PARAMETERS :\n");
	if (maxbyte == 0)
		Ctape_dmpmsg (MSG_OUT, " DUMP - ALL BYTES\n");
	else
		Ctape_dmpmsg (MSG_OUT, " DUMP - MAX. NB OF BYTES: %d\n", maxbyte);
	Ctape_dmpmsg (MSG_OUT, " DUMP - FROM BLOCK: %d\n", fromblock);
	if (toblock == 0)
		Ctape_dmpmsg (MSG_OUT, " DUMP - TO LAST BLOCK\n");
	else
		Ctape_dmpmsg (MSG_OUT, " DUMP - TO BLOCK: %d\n", toblock);
	if (fromfile > 0)
		Ctape_dmpmsg (MSG_OUT, " DUMP - FROM FILE %d\n", fromfile);
	if (maxfile == 0)
		Ctape_dmpmsg (MSG_OUT, " DUMP - ALL FILES\n");
	else
		Ctape_dmpmsg (MSG_OUT, " DUMP - MAX. NB OF FILES: %d\n", maxfile);
	Ctape_dmpmsg (MSG_OUT, " DUMP - ASCII INTERPRETATION\n");
	Ctape_dmpmsg (MSG_OUT, " DUMP -\n");

	/* open path */

	if (infd < 0 && (infd = open (path, O_RDONLY)) < 0) {
		Ctape_dmpmsg (MSG_ERR, " DUMP ! ERROR OPENING TAPE FILE: %s\n",
			strerror(errno));
		serrno = errno;
		return (-1);
	}

	(void) clear_compression_stats (infd, path, devtype);

	/* save dump parameters */

	strcpy (dmpparm.vid, vid);
	strcpy (dmpparm.devtype, devtype);
	dmpparm.maxblksize = maxblksize;
	dmpparm.fromblock = fromblock;
	dmpparm.toblock = toblock;
	dmpparm.maxbyte = maxbyte;
	dmpparm.fromfile = fromfile;
	dmpparm.maxfile = maxfile;
	dmpparm.code = code;
	dmpparm.flags = flags;
	return (0);
}

int ebc_asc(char *p,
            char *q,
            int n)
{
	int i;
	static char e2atab[256] = {

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	' ','.','.','.','.','.','.','.','.','.','.','.','<','(','+','|',

	'&','.','.','.','.','.','.','.','.','.','!','$','*',')',';','^',

	'-','/','.','.','.','.','.','.','.','.','.',',','%','_','>','?',

	'.','.','.','.','.','.','.','.','.','`',':','#','@','\'','=','"',

	'.','a','b','c','d','e','f','g','h','i','.','.','.','.','.','.',

	'.','j','k','l','m','n','o','p','q','r','.','.','.','.','.','.',

	'.','~','s','t','u','v','w','x','y','z','.','.','.','[','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.',']','.','.',

	'{','A','B','C','D','E','F','G','H','I','.','.','.','.','.','.',

	'}','J','K','L','M','N','O','P','Q','R','.','.','.','.','.','.',

	'\\','.','S','T','U','V','W','X','Y','Z','.','.','.','.','.','.',

	'0','1','2','3','4','5','6','7','8','9','.','.','.','.','.','.',
	};

	for (i = 0; i < n; i++)
		*q++ = e2atab[*p++ & 0xff];
        return (0);
}

int asc_asc(char *p,
            char *q,
            int n)
{
	int i;
	static char a2atab[256] = {

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	' ','!','"','#','$','%','&','\'','(',')','*','+',',','-','.','/',

	'0','1','2','3','4','5','6','7','8','9',':',';','<','=','>','?',

	'@','A','B','C','D','E','F','G','H','I','J','K','L','M','N','O',

	'P','Q','R','S','T','U','V','W','X','Y','Z','[','\\',']','^','_',

	'`','a','b','c','d','e','f','g','h','i','j','k','l','m','n','o',

	'p','q','r','s','t','u','v','w','x','y','z','{','|','}','~','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',

	'.','.','.','.','.','.','.','.','.','.','.','.','.','.','.','.',
	};

	for (i = 0; i < n; i++)
		*q++ = a2atab[*p++ & 0xff];
        return (0);
}


int dumpblock (char *buffer,
               int nbytes)
{
	char bufftr[33];
	int i, k, l;
	int iad;
	char *p;

	Ctape_dmpmsg (MSG_OUT, " BLOCK NUMBER %d\n", irec);
	for (iad = 0, p = buffer; iad < nbytes; iad += 32) {
		if (dmpparm.maxbyte != 0 && iad >= dmpparm.maxbyte) break;
		l = iad + 31;
		if (l >= nbytes) l = nbytes - 1;
		k = l - iad + 1;
		if (dmpparm.code == DMP_EBC)
			ebc_asc (p, bufftr, k);
		else
			asc_asc (p, bufftr, k);
		bufftr[k] = '\0';
		Ctape_dmpmsg (MSG_OUT, " %.8x   ", iad);
		for (i = 0; i < k; i++) {
			if (i % 8 == 0) putchar (' ');
			Ctape_dmpmsg (MSG_OUT, "%.2x", *(p++) & 0xff);
		}
		for (i = k; i < 32; i++)
			putchar (' ');
		Ctape_dmpmsg (MSG_OUT, "    *%s*\n", bufftr);
	}
        return (0);
}

int printlabel1 (char *label)
{
	Ctape_dmpmsg (MSG_OUT, " FILE ID:                    %.17s\n", label + 4);
	Ctape_dmpmsg (MSG_OUT, " VOLUME SEQNO:               %.4s\n", label + 27);
	Ctape_dmpmsg (MSG_OUT, " FILE SEQNO:                 %.4s\n", label + 31);
	Ctape_dmpmsg (MSG_OUT, " CREATION DATE:              %.6s\n", label + 41);
	Ctape_dmpmsg (MSG_OUT, " EXPIRATION DATE:            %.6s\n", label + 47);
	Ctape_dmpmsg (MSG_OUT, " BLOCK COUNT:                %.6s\n", label + 54);

        return (0);
}

int printlabel2 (char *label)
{
	Ctape_dmpmsg (MSG_OUT, " RECORD FORMAT:              %.1s\n", label + 4);
	Ctape_dmpmsg (MSG_OUT, " BLOCK LENGTH:               %.5s\n", label + 5);
	Ctape_dmpmsg (MSG_OUT, " RECORD LENGTH:              %.5s\n", label + 10);
	Ctape_dmpmsg (MSG_OUT, " DENSITY:                    %.1s\n", label + 15);
	Ctape_dmpmsg (MSG_OUT, " DATA RECORDING:             %.1s\n", label + 34);
	Ctape_dmpmsg (MSG_OUT, " BLOCKING ATTRIBUTE:         %.1s\n", label + 38);

        return (0);
}

int printulabel1 (char *label)
{
	Ctape_dmpmsg (MSG_OUT, " FILE SEQNO:                 %.10s\n", label + 4);
	Ctape_dmpmsg (MSG_OUT, " BLOCK LENGTH:               %.10s\n", label + 14);
	Ctape_dmpmsg (MSG_OUT, " RECORD LENGTH:              %.10s\n", label + 24);
	Ctape_dmpmsg (MSG_OUT, " SITE:                       %.8s\n", label + 34);
	Ctape_dmpmsg (MSG_OUT, " TAPE MOVER HOSTNAME:        %.10s\n", label + 42);
	Ctape_dmpmsg (MSG_OUT, " DRIVE MANUFACTURER:         %.8s\n", label + 52);
	Ctape_dmpmsg (MSG_OUT, " DRIVE MODEL:                %.8s\n", label + 60);
	Ctape_dmpmsg (MSG_OUT, " DRIVE SERIAL NUMBER:        %.12s\n", label + 68);

        return (0);
}

int islabel(char *label)
{
	if (strncmp (label, "HDR1", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n HEADER LABEL 1:\n");
		printlabel1 (label);
		return (1);
	} else if (strncmp (label, "HDR2", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n HEADER LABEL 2:\n");
		printlabel2 (label);
		return (2);
	} else if (strncmp (label, "HDR3", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n HEADER LABEL 3:             %.76s\n", label+4);
		return (3);
	} else if (strncmp (label, "HDR4", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n HEADER LABEL 4:             %.76s\n", label+4);
		return (4);
	} else if (strncmp (label, "UHL1", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n USER HEADER LABEL 1:\n");
		printulabel1 (label);
		return (5);
	} else if (strncmp (label, "EOF1", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n TRAILER LABEL 1:\n");
		printlabel1 (label);
		return (-1);
	} else if (strncmp (label, "EOF2", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n TRAILER LABEL 2:\n");
		printlabel2 (label);
		return (-2);
	} else if (strncmp (label, "EOF3", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n TRAILER LABEL 3:            %.76s\n", label+4);
		return (-3);
	} else if (strncmp (label, "EOF4", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n TRAILER LABEL 4:            %.76s\n", label+4);
		return (-4);
	} else if (strncmp (label, "EOV1", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n END OF VOLUME LABEL 1:\n");
		printlabel1 (label);
		return (-1);
	} else if (strncmp (label, "EOV2", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n END OF VOLUME LABEL 2:\n");
		printlabel2 (label);
		return (-2);
	} else if (strncmp (label, "UTL1", 4) == 0) {
		Ctape_dmpmsg (MSG_OUT, "\n USER TRAILER LABEL 1:\n");
		printulabel1 (label);
		return (-5);
	}
	return (0);
}

int report_comp_stats (int infd,
                       char *path,
                       char *devtype)
{
	COMPRESSION_STATS compstats;

	memset (&compstats, 0, sizeof(compstats));
	if (get_compression_stats (infd, path, devtype, &compstats) == 0) {
		if (compstats.from_tape)
			Ctape_dmpmsg (MSG_OUT, " ***** COMPRESSION RATE ON TAPE: %8.2f\n",
			    (float)compstats.to_host / (float)compstats.from_tape);
	}
	return (0);
}

int Ctape_dmpend()
{
	if (infd >= 0)
		close (infd);
	infd = -1;
	return (0);
}


int Ctape_dmpfil(char *path,
                 char *lbltype,
                 int *blksize,
                 char *fid,
                 int *fsec,
                 int *fseq,
                 int *lrecl,
                 char *recfm,
                 u_signed64 *Size)
{
	int avg_block_length;
	unsigned char blockid[4];
	int errcat = 0;
	u_signed64 filesize;
	int goodrec = 0;
	char label[81];
	static int volHasAulFormat = 0;
	int max_block_length = 0;
	int min_block_length = 0;
	char *msgaddr;
	int nbytes;
	static int nerr = 0;
	static int nfile = 0;
	char *p;
	static int qbov = 1;
	int qlab = 0;
	u_signed64 sum_block_length = 0;
	static float tape_used = 0;
	char tmpbuf[21];

	filesize = 0;
	if (read_pos (infd, path, blockid) == 0) {
		Ctape_dmpmsg (MSG_OUT, " BLOCKID:                    %02x%02x%02x%02x\n",
                      blockid[0], blockid[1], blockid[2], blockid[3]);
	} else {
		Ctape_dmpmsg(MSG_OUT, " BLOCKID: Retrieve error\n");
	}
	while (1) {
		nbytes = read (infd, buffer, dmpparm.maxblksize);
		if (nbytes > 0) {		/* record found */
			irec++;
			tape_used = tape_used + (float) nbytes;
			goodrec++;
			if (min_block_length == 0) min_block_length = nbytes;
			if (nbytes > max_block_length)
				max_block_length = nbytes;
			else if (nbytes < min_block_length)
				min_block_length = nbytes;
			sum_block_length += nbytes;

			if (qbov) {	/* beginning of tape */
				qbov = 0;
				if (nbytes == 80) {
					asc_asc (buffer, label, 80);
					if (strncmp (label, "VOL1", 4) == 0)
						volHasAulFormat = 1; 
				}
				if (volHasAulFormat) {
					Ctape_dmpmsg (MSG_OUT, "               %s HAS AN ASCII LABEL\n",
					    dmpparm.vid);
					Ctape_dmpmsg (MSG_OUT, "\n VOLUME LABEL :\n");
					Ctape_dmpmsg (MSG_OUT, " VOLUME SERIAL NUMBER:       %.6s\n",
					    label+4);
					Ctape_dmpmsg (MSG_OUT, " OWNER IDENTIFIER:           %.14s\n",
					    label+37);
					if (lbltype) {
						strcpy (lbltype, "aul");
					}
				} else {
					Ctape_dmpmsg (MSG_OUT, "               %s IS UNLABELLED\n",
					    dmpparm.vid);
					if (lbltype) {
                                                strcpy (lbltype, "nl");
					}
				}
				if (dmpparm.fromfile > 1) {
					const int n = volHasAulFormat ? (dmpparm.fromfile - 1)*3 : (dmpparm.fromfile - 1);
					goodrec = 0;
					Ctape_dmpmsg (MSG_OUT, "\n DUMP - SKIPPING TO FILE %d\n",
					    dmpparm.fromfile);
					if (skiptpff (infd, path, n) < 0) {
						Ctape_dmpmsg (MSG_ERR, " DUMP ! SKIP ERROR\n");
						errcat = EIO;
						break;
					}
					irec = 0;
					max_block_length = 0;
					min_block_length = 0;
					sum_block_length = 0;
					continue;
				}
				if (volHasAulFormat) {
					continue;
				}
			} else if (irec < 6 && nbytes == 80) {
				asc_asc (buffer, label, 80);
				qlab = islabel (label);
				switch (qlab) {
				case 0:	/* not a label */
					break;
				case 1:	/* HDR1 */
					if (fid) {
						strncpy (fid, label + 4, 17);
						*(fid+17) = '\0';
						if ((p = strchr (fid, ' '))) *p = '\0';
					}
					if (fsec) {
						sscanf (label+27, "%4d", fsec);
					}
					if (fseq) {
						sscanf (label+31, "%4d", fseq);
					}
					continue;
				case 2:	/* HDR2 */
					if (recfm) {
						memset (recfm, 0, 4);
						*recfm = label[4];
					}
					if (blksize) {
						sscanf (label+5, "%5d", blksize);
					}
					if (lrecl) {
						sscanf (label+10, "%5d", lrecl);
					}
					continue;
				case 5: /* UHL1 */
					if (fseq) {
						sscanf (label+4, "%10d", fseq);
					}
					if (blksize) {
						sscanf (label+14, "%10d", blksize);
					}
					if (lrecl) {
						sscanf (label+24, "%10d", lrecl);
					}
					continue;
				default:
					continue;
				}
			}
			qlab = 0;
			filesize += nbytes;
			if (irec >= dmpparm.fromblock &&
				(dmpparm.toblock == 0 || irec <= dmpparm.toblock)) {
				dumpblock (buffer, nbytes);
				Ctape_dmpmsg (MSG_OUT, " BLOCK NUMBER %d  LENGTH = %d BYTES\n",
					irec, nbytes);
			}

		} else if (nbytes == 0) {	/* tape mark found */
#if defined(sun) || defined(linux)
			if (gettperror (infd, path, &msgaddr) == ETBLANK) {
				if (!qbov && irec == 0) break;
				fflush (stdout);
				Ctape_dmpmsg (MSG_ERR, " DUMP ! READ ERROR: %s (BLOCK # %d)\n",
					"Blank check", irec+1);
				break;
			}
#endif
			if (qbov) {	/* beginning of tape */
				Ctape_dmpmsg (MSG_OUT, "               %s IS UNLABELLED\n", dmpparm.vid);
				if (lbltype) {
					strcpy (lbltype, "nl");
				}
			}
			if (qlab) {
				Ctape_dmpmsg (MSG_OUT, " ***** TAPE MARK READ *****      END OF LABEL GROUP\n");
				Ctape_dmpmsg (MSG_OUT, " *********************************************************************************************************************\n");
				if (dmpparm.maxfile != 0 && nfile >= dmpparm.maxfile) {
					(void) report_comp_stats (infd, path, dmpparm.devtype);
					Ctape_dmpmsg (MSG_OUT, " DUMP - DUMPING PROGRAM COMPLETE.\n");
					close (infd);
					return (1);
				}
			} else {
				nfile++;
				Ctape_dmpmsg (MSG_OUT, " ***** TAPE MARK READ *****      FILE %d CONTAINED %d BLOCKS.\n",
					nfile, irec);
				if (Size)
					*Size = filesize;
				if (goodrec != 0) {
					avg_block_length = sum_block_length / goodrec;
					Ctape_dmpmsg (MSG_OUT, "                                 FILE SIZE WAS %s BYTES\n",
						u64tostr (filesize, tmpbuf, 0));
					Ctape_dmpmsg (MSG_OUT, "                                 MAXIMUM BLOCK LENGTH WAS %d BYTES\n",
						max_block_length);
					Ctape_dmpmsg (MSG_OUT, "                                 MINIMUM BLOCK LENGTH WAS %d BYTES\n",
						min_block_length);
					Ctape_dmpmsg (MSG_OUT, "                                 AVERAGE BLOCK LENGTH WAS %d BYTES\n",
						avg_block_length);
					Ctape_dmpmsg (MSG_OUT, " *********************************************************************************************************************\n");
					if (!volHasAulFormat && dmpparm.maxfile != 0 &&
					    nfile >= dmpparm.maxfile) {
						(void) report_comp_stats (infd, path, dmpparm.devtype);
						Ctape_dmpmsg (MSG_OUT, " DUMP - DUMPING PROGRAM COMPLETE.\n");
						close (infd);
						return (1);
					}
				} else if ((dmpparm.flags & IGNOREEOI) == 0 && !qbov) {
					Ctape_dmpmsg (MSG_OUT, " ***** DOUBLE TAPE MARK READ *****\n");
					break;
				}
			}
			irec = 0;
			qbov = 0;
			if (!volHasAulFormat || qlab < 0) return (0);
			goodrec = 0;
			max_block_length = 0;
			min_block_length = 0;
			qlab = 0;
			sum_block_length = 0;
		} else {
			errcat = 0;
			if (errno == ENOMEM) {
				msgaddr = "Large block";
				errcat = ENOMEM;
			} else if (errno == EIO)
				errcat = gettperror (infd, path, &msgaddr);
			else
					msgaddr = (char *) strerror(errno);
			if (errcat == ETBLANK && !qbov && irec == 0) break;
			irec++;
			fflush (stdout);
			Ctape_dmpmsg (MSG_ERR, " DUMP ! READ ERROR: %s (BLOCK # %d)\n",
				msgaddr, irec);
			if (errcat == ETBLANK) break;
			if (++nerr >= 10) {
				Ctape_dmpmsg (MSG_ERR, " DUMP ! TOO MANY READ ERRORS.\n");
				break;
			}
		}
	}
	if (irec) {
		if (goodrec) {
			avg_block_length = sum_block_length / goodrec;
			Ctape_dmpmsg (MSG_OUT, "                                 MAXIMUM BLOCK LENGTH WAS %d BYTES\n",
				max_block_length);
			Ctape_dmpmsg (MSG_OUT, "                                 MINIMUM BLOCK LENGTH WAS %d BYTES\n",
				min_block_length);
			Ctape_dmpmsg (MSG_OUT, "                                 AVERAGE BLOCK LENGTH WAS %d BYTES\n",
				avg_block_length);
			Ctape_dmpmsg (MSG_OUT, " *********************************************************************************************************************\n");
		}
		Ctape_dmpmsg (MSG_OUT, " DUMP - DUMP ABORTED\n");
		serrno = (errcat > 0) ? errcat : EIO;
		close (infd);
		return (-1);
	}

	Ctape_dmpmsg (MSG_OUT, "\n ***** THE RECORDED DATA OCCUPIED %lu bytes OF A CARTRIDGE  *****\n",
		(unsigned long)tape_used) ;

	(void) report_comp_stats (infd, path, dmpparm.devtype);
	Ctape_dmpmsg (MSG_OUT, " DUMP - DUMPING PROGRAM COMPLETE.\n");
	close (infd);
	return (1);
}
